File: experiments-local/experiment-mu-kimi3-brazil/runs/agent=opencode_language=typescript_model=openrouter/moonshotai/kimi-k3_tooling=none/rep3/src/queries.ts
/**
 * Query engine: answers the five question categories from the spec
 * (matches, teams, players, competitions, statistics) on top of the
 * knowledge-graph indexes.
 */

import type {
  CompetitionLabel,
  Match,
  Player,
  StandingRow,
  TeamRecord,
} from "./types.js";
import type { BuiltGraph } from "./graph.js";
import {
  isBrazilianTeamKey,
  loose,
  resolveCompetition,
  resolveTeamQuery,
  teamDisplayName,
} from "./normalize.js";

export class SoccerQueries {
  constructor(private readonly built: BuiltGraph) {}

  private get matches(): Match[] {
    // All matches, already sorted by date asc.
    return this.allMatches;
  }

  private allMatches: Match[] = [];

  /** Bind the flat match list (kept separate from graph for iteration). */
  bindMatches(matches: Match[]): this {
    this.allMatches = matches;
    return this;
  }

  /* ---------------------------------------------------------------- */
  /* 1. Match queries                                                   */
  /* ---------------------------------------------------------------- */

  findMatches(filter: {
    team?: string;
    opponent?: string;
    competition?: string;
    season?: number;
    from?: string;
    to?: string;
    venue?: "home" | "away" | "any";
    round?: string;
    stage?: string;
    limit?: number;
  }): Match[] {
    const teamKey = filter.team ? resolveTeamQuery(filter.team) : null;
    const oppKey = filter.opponent ? resolveTeamQuery(filter.opponent) : null;
    const comp = filter.competition
      ? resolveCompetition(filter.competition)
      : null;
    const venue = filter.venue ?? "any";
    const limit = Math.min(Math.max(filter.limit ?? 50, 1), 500);

    let pool: Match[];
    if (teamKey && this.built.indexes.matchesByTeam.has(teamKey)) {
      pool = this.built.indexes.matchesByTeam.get(teamKey)!;
    } else if (comp && this.built.indexes.matchesByCompetition.has(comp)) {
      pool = this.built.indexes.matchesByCompetition.get(comp)!;
    } else {
      pool = this.matches;
    }

    const out: Match[] = [];
    for (const m of pool) {
      if (teamKey) {
        const isHome = m.homeTeam.key === teamKey;
        const isAway = m.awayTeam.key === teamKey;
        if (venue === "home" && !isHome) continue;
        if (venue === "away" && !isAway) continue;
        if (!isHome && !isAway) continue;
      }
      if (oppKey && m.homeTeam.key !== oppKey && m.awayTeam.key !== oppKey) continue;
      if (comp && m.competition !== comp) continue;
      if (filter.season != null && m.season !== filter.season) continue;
      if (filter.from && m.date < filter.from) continue;
      if (filter.to && m.date > filter.to) continue;
      if (filter.round && m.round !== filter.round) continue;
      if (filter.stage && loose(m.stage ?? "") !== loose(filter.stage)) continue;
      out.push(m);
      if (out.length >= limit) break;
    }
    return out;
  }

  /* ---------------------------------------------------------------- */
  /* Head-to-head                                                       */
  /* ---------------------------------------------------------------- */

  headToHead(teamA: string, teamB: string, competition?: string) {
    const keyA = resolveTeamQuery(teamA);
    const keyB = resolveTeamQuery(teamB);
    const comp = competition ? resolveCompetition(competition) : null;
    const pool = this.built.indexes.matchesByTeam.get(keyA) ?? [];
    const matches = pool.filter(
      (m) =>
        (m.homeTeam.key === keyB || m.awayTeam.key === keyB) &&
        (!comp || m.competition === comp),
    );
    let winsA = 0, winsB = 0, draws = 0, goalsA = 0, goalsB = 0;
    for (const m of matches) {
      if (m.homeGoals == null || m.awayGoals == null) continue;
      const aHome = m.homeTeam.key === keyA;
      const gA = aHome ? m.homeGoals : m.awayGoals;
      const gB = aHome ? m.awayGoals : m.homeGoals;
      goalsA += gA;
      goalsB += gB;
      if (gA > gB) winsA++;
      else if (gA < gB) winsB++;
      else draws++;
    }
    return {
      teamA: teamDisplayName(keyA),
      teamB: teamDisplayName(keyB),
      matches,
      summary: {
        total: matches.length,
        winsA,
        draws,
        winsB,
        goalsA,
        goalsB,
      },
    };
  }

  /* ---------------------------------------------------------------- */
  /* 2. Team queries                                                    */
  /* ---------------------------------------------------------------- */

  teamRecord(
    team: string,
    opts: { season?: number; competition?: string; venue?: "home" | "away" | "any" } = {},
  ): TeamRecord & { teamKey: string } {
    return this.recordForKey(resolveTeamQuery(team), opts);
  }

  private recordForKey(
    key: string,
    opts: { season?: number; competition?: string; venue?: "home" | "away" | "any" } = {},
  ): TeamRecord & { teamKey: string } {
    const comp = opts.competition ? resolveCompetition(opts.competition) : null;
    const venue = opts.venue ?? "any";
    const pool = this.built.indexes.matchesByTeam.get(key) ?? [];
    let wins = 0, draws = 0, losses = 0, gf = 0, ga = 0, played = 0;
    for (const m of pool) {
      const isHome = m.homeTeam.key === key;
      if (venue === "home" && !isHome) continue;
      if (venue === "away" && isHome) continue;
      if (comp && m.competition !== comp) continue;
      if (opts.season != null && m.season !== opts.season) continue;
      if (m.homeGoals == null || m.awayGoals == null) continue;
      played++;
      const scored = isHome ? m.homeGoals : m.awayGoals;
      const conceded = isHome ? m.awayGoals : m.homeGoals;
      gf += scored;
      ga += conceded;
      if (scored > conceded) wins++;
      else if (scored < conceded) losses++;
      else draws++;
    }
    return {
      team: teamDisplayName(key),
      teamKey: key,
      matches: played,
      wins,
      draws,
      losses,
      goalsFor: gf,
      goalsAgainst: ga,
      goalDifference: gf - ga,
      winRate: played ? wins / played : 0,
    };
  }

  /** Competitions (with seasons + match counts) a team has played in. */
  teamCompetitions(team: string) {
    const key = resolveTeamQuery(team);
    const pool = this.built.indexes.matchesByTeam.get(key) ?? [];
    const byComp = new Map<CompetitionLabel, { seasons: Set<number>; matches: number }>();
    for (const m of pool) {
      const entry = byComp.get(m.competition) ?? { seasons: new Set<number>(), matches: 0 };
      entry.matches++;
      if (m.season != null) entry.seasons.add(m.season);
      byComp.set(m.competition, entry);
    }
    return {
      team: teamDisplayName(key),
      teamKey: key,
      competitions: [...byComp.entries()].map(([competition, e]) => ({
        competition,
        matches: e.matches,
        seasons: [...e.seasons].sort((a, b) => a - b),
      })),
    };
  }

  /* ---------------------------------------------------------------- */
  /* 3. Player queries                                                  */
  /* ---------------------------------------------------------------- */

  searchPlayers(filter: {
    name?: string;
    nationality?: string;
    club?: string;
    position?: string;
    minOverall?: number;
    brazilianClubsOnly?: boolean;
    limit?: number;
    sortByOverall?: boolean;
  }): Player[] {
    const limit = Math.min(Math.max(filter.limit ?? 20, 1), 200);
    let pool: Player[] = this.allPlayers;

    if (filter.nationality) {
      const nat = loose(filter.nationality);
      pool = pool.filter((p) => loose(p.nationality) === nat);
    }
    if (filter.club) {
      const clubKey = resolveTeamQuery(filter.club);
      const clubLoose = loose(filter.club);
      pool = pool.filter((p) => {
        if (!p.club) return false;
        if (p.clubKey && p.clubKey === clubKey) return true;
        return loose(p.club).includes(clubLoose);
      });
    }
    if (filter.brazilianClubsOnly) {
      pool = pool.filter((p) => p.clubKey != null && isBrazilianTeamKey(p.clubKey));
    }
    if (filter.position) {
      const pos = filter.position.toUpperCase();
      const groups: Record<string, string[]> = {
        FORWARD: ["ST", "CF", "LW", "RW", "LF", "RF", "LS", "RS"],
        MIDFIELDER: ["CAM", "CM", "CDM", "LM", "RM", "LAM", "RAM", "LCM", "RCM", "LDM", "RDM"],
        DEFENDER: ["CB", "LB", "RB", "LCB", "RCB", "LWB", "RWB"],
        GOALKEEPER: ["GK"],
      };
      const wanted = groups[pos] ?? [pos];
      pool = pool.filter((p) => p.position != null && wanted.includes(p.position));
    }
    if (filter.minOverall != null) {
      pool = pool.filter((p) => (p.overall ?? 0) >= filter.minOverall!);
    }
    if (filter.name) {
      const q = loose(filter.name);
      pool = pool.filter((p) => loose(p.name).includes(q));
    }
    if (filter.sortByOverall !== false) {
      pool = [...pool].sort(
        (a, b) => (b.overall ?? 0) - (a.overall ?? 0) || a.name.localeCompare(b.name),
      );
    }
    return pool.slice(0, limit);
  }

  /** Count + average rating of a nationality's players per club. */
  playersByClubSummary(nationality: string, brazilianClubsOnly = true) {
    const nat = loose(nationality);
    const pool = this.allPlayers.filter(
      (p) =>
        loose(p.nationality) === nat &&
        (!brazilianClubsOnly || (p.clubKey != null && isBrazilianTeamKey(p.clubKey))),
    );
    const byClub = new Map<string, { count: number; totalOverall: number; club: string }>();
    for (const p of pool) {
      const key = p.clubKey ?? loose(p.club ?? "unknown");
      const entry = byClub.get(key) ?? {
        count: 0,
        totalOverall: 0,
        club: p.clubKey ? teamDisplayName(p.clubKey) : (p.club ?? "Unknown"),
      };
      entry.count++;
      entry.totalOverall += p.overall ?? 0;
      byClub.set(key, entry);
    }
    return [...byClub.values()]
      .map((e) => ({
        club: e.club,
        players: e.count,
        averageOverall: e.count ? Math.round((e.totalOverall / e.count) * 10) / 10 : 0,
      }))
      .sort((a, b) => b.players - a.players || b.averageOverall - a.averageOverall);
  }

  /* ---------------------------------------------------------------- */
  /* 4. Competition queries                                             */
  /* ---------------------------------------------------------------- */

  standings(season: number, competition = "Brasileirão Série A"): StandingRow[] {
    const comp = resolveCompetition(competition) ?? "Brasileirão Série A";
    const pool = (this.built.indexes.matchesByCompetition.get(comp) ?? []).filter(
      (m) => m.season === season && m.homeGoals != null && m.awayGoals != null,
    );
    const table = new Map<string, StandingRow>();
    const ensure = (key: string): StandingRow => {
      let row = table.get(key);
      if (!row) {
        row = {
          rank: 0,
          team: teamDisplayName(key),
          matches: 0,
          wins: 0,
          draws: 0,
          losses: 0,
          goalsFor: 0,
          goalsAgainst: 0,
          goalDifference: 0,
          winRate: 0,
          points: 0,
        };
        table.set(key, row);
      }
      return row;
    };
    for (const m of pool) {
      const home = ensure(m.homeTeam.key);
      const away = ensure(m.awayTeam.key);
      const hg = m.homeGoals!;
      const ag = m.awayGoals!;
      home.matches++;
      away.matches++;
      home.goalsFor += hg;
      home.goalsAgainst += ag;
      away.goalsFor += ag;
      away.goalsAgainst += hg;
      if (hg > ag) {
        home.wins++;
        away.losses++;
        home.points += 3;
      } else if (hg < ag) {
        away.wins++;
        home.losses++;
        away.points += 3;
      } else {
        home.draws++;
        away.draws++;
        home.points += 1;
        away.points += 1;
      }
    }
    const rows = [...table.values()];
    for (const r of rows) {
      r.goalDifference = r.goalsFor - r.goalsAgainst;
      r.winRate = r.matches ? r.wins / r.matches : 0;
    }
    rows.sort(
      (a, b) =>
        b.points - a.points ||
        b.wins - a.wins ||
        b.goalDifference - a.goalDifference ||
        b.goalsFor - a.goalsFor ||
        a.team.localeCompare(b.team),
    );
    rows.forEach((r, i) => (r.rank = i + 1));
    if (rows.length > 0) rows[0].note = "Champion";
    if (comp === "Brasileirão Série A" && rows.length >= 20) {
      for (const r of rows.slice(-4)) r.note = r.note ? `${r.note}; Relegated` : "Relegated";
    }
    return rows;
  }

  /** Copa do Brasil "finals": matches in the last (max) round of each season. */
  cupFinals(season?: number): Match[] {
    const pool = this.built.indexes.matchesByCompetition.get("Copa do Brasil") ?? [];
    const bySeason = new Map<number, Match[]>();
    for (const m of pool) {
      if (m.season == null) continue;
      if (season != null && m.season !== season) continue;
      const arr = bySeason.get(m.season) ?? [];
      arr.push(m);
      bySeason.set(m.season, arr);
    }
    const finals: Match[] = [];
    for (const [s, matches] of [...bySeason.entries()].sort((a, b) => a[0] - b[0])) {
      let maxRound = -1;
      for (const m of matches) {
        const r = parseInt(m.round ?? "", 10);
        if (Number.isFinite(r) && r > maxRound) maxRound = r;
      }
      finals.push(...matches.filter((m) => parseInt(m.round ?? "", 10) === maxRound));
    }
    return finals;
  }

  /** Seasons available per competition. */
  competitionSeasons(): Record<string, number[]> {
    const out: Record<string, Set<number>> = {};
    for (const [comp, matches] of this.built.indexes.matchesByCompetition) {
      out[comp] = out[comp] ?? new Set();
      for (const m of matches) if (m.season != null) out[comp].add(m.season);
    }
    return Object.fromEntries(
      Object.entries(out).map(([k, v]) => [k, [...v].sort((a, b) => a - b)]),
    );
  }

  /* ---------------------------------------------------------------- */
  /* 5. Statistical analysis                                            */
  /* ---------------------------------------------------------------- */

  competitionStats(opts: { competition?: string; season?: number } = {}) {
    const comp = opts.competition ? resolveCompetition(opts.competition) : null;
    let played = 0, goals = 0, homeWins = 0, awayWins = 0, draws = 0;
    for (const m of this.matches) {
      if (comp && m.competition !== comp) continue;
      if (opts.season != null && m.season !== opts.season) continue;
      if (m.homeGoals == null || m.awayGoals == null) continue;
      played++;
      goals += m.homeGoals + m.awayGoals;
      if (m.homeGoals > m.awayGoals) homeWins++;
      else if (m.homeGoals < m.awayGoals) awayWins++;
      else draws++;
    }
    return {
      competition: comp ?? "All competitions",
      season: opts.season ?? "all",
      matchesPlayed: played,
      totalGoals: goals,
      averageGoalsPerMatch: played ? Math.round((goals / played) * 100) / 100 : 0,
      homeWins,
      draws,
      awayWins,
      homeWinRate: played ? Math.round((homeWins / played) * 1000) / 10 : 0,
      drawRate: played ? Math.round((draws / played) * 1000) / 10 : 0,
      awayWinRate: played ? Math.round((awayWins / played) * 1000) / 10 : 0,
    };
  }

  biggestWins(opts: { competition?: string; season?: number; limit?: number } = {}) {
    const comp = opts.competition ? resolveCompetition(opts.competition) : null;
    const limit = Math.min(Math.max(opts.limit ?? 10, 1), 100);
    const played = this.matches.filter(
      (m) =>
        m.homeGoals != null &&
        m.awayGoals != null &&
        (!comp || m.competition === comp) &&
        (opts.season == null || m.season === opts.season),
    );
    const withMargin = played.map((m) => ({
      match: m,
      margin: Math.abs(m.homeGoals! - m.awayGoals!),
      totalGoals: m.homeGoals! + m.awayGoals!,
    }));
    withMargin.sort((a, b) => b.margin - a.margin || b.totalGoals - a.totalGoals || a.match.date.localeCompare(b.match.date));
    return withMargin.slice(0, limit).map((w) => w.match);
  }

  /** Best home records (by win rate, min 10 home games). */
  bestHomeRecords(opts: { competition?: string; season?: number; limit?: number; minMatches?: number } = {}) {
    return this.bestVenueRecords("home", opts);
  }

  /** Best away records (by win rate, min 10 away games). */
  bestAwayRecords(opts: { competition?: string; season?: number; limit?: number; minMatches?: number } = {}) {
    return this.bestVenueRecords("away", opts);
  }

  private bestVenueRecords(
    venue: "home" | "away",
    opts: { competition?: string; season?: number; limit?: number; minMatches?: number },
  ): TeamRecord[] {
    const limit = Math.min(Math.max(opts.limit ?? 10, 1), 100);
    const minMatches = opts.minMatches ?? 10;
    const records: TeamRecord[] = [];
    for (const key of this.built.indexes.teams.keys()) {
      const rec = this.recordForKey(key, {
        season: opts.season,
        competition: opts.competition,
        venue,
      });
      if (rec.matches >= minMatches) records.push(rec);
    }
    records.sort(
      (a, b) =>
        b.winRate - a.winRate || b.wins - a.wins || b.goalDifference - a.goalDifference,
    );
    return records.slice(0, limit);
  }

  /** Top-scoring teams (goals for) in a season/competition. */
  topScoringTeams(opts: { season?: number; competition?: string; limit?: number } = {}) {
    const limit = Math.min(Math.max(opts.limit ?? 10, 1), 100);
    const records: TeamRecord[] = [];
    for (const key of this.built.indexes.teams.keys()) {
      const rec = this.recordForKey(key, {
        season: opts.season,
        competition: opts.competition,
      });
      if (rec.matches > 0) records.push(rec);
    }
    records.sort((a, b) => b.goalsFor - a.goalsFor || b.goalDifference - a.goalDifference);
    return records.slice(0, limit);
  }

  /* ---------------------------------------------------------------- */
  /* Dataset overview                                                   */
  /* ---------------------------------------------------------------- */

  private allPlayers: Player[] = [];

  bindPlayers(players: Player[]): this {
    this.allPlayers = players;
    return this;
  }

  overview(sourceRowCounts: Record<string, number>, duplicateCounts: Record<string, number>) {
    const seasons = new Set<number>();
    for (const m of this.matches) if (m.season != null) seasons.add(m.season);
    return {
      sources: sourceRowCounts,
      duplicatesMerged: duplicateCounts,
      uniqueMatches: this.matches.length,
      teams: this.built.indexes.teams.size,
      players: this.allPlayers.length,
      brazilianPlayers: this.allPlayers.filter((p) => p.nationality === "Brazil").length,
      competitions: this.competitionSeasons(),
      seasonRange: seasons.size
          ? [Math.min(...seasons), Math.max(...seasons)]
          : null,
    };
  }
}

/** Convenience: a fully wired query engine. */
export function createQueries(
  built: BuiltGraph,
  matches: Match[],
  players: Player[],
): SoccerQueries {
  return new SoccerQueries(built).bindMatches(matches).bindPlayers(players);
}

File: experiments-local/experiment-mu-glm53-thinking/runs/agent=oc-think-off_language=python_model=openrouter/z-ai/glm-5.3_tooling=none/rep3/brazilian_soccer/queries.py
"""Query layer: statistics, standings, head-to-head, and formatting.

All high-level question answering used by the MCP tools lives here so it
can be tested independently of the MCP transport.
"""

from __future__ import annotations

from collections import defaultdict
from datetime import datetime
from typing import Optional

from .loader import Match, SoccerData, name_matches, normalize_name


def _fmt_score(m: Match) -> str:
    hg = "?" if m.home_goal is None else m.home_goal
    ag = "?" if m.away_goal is None else m.away_goal
    return f"{hg}-{ag}"


def _match_line(m: Match) -> str:
    date = m.date.strftime("%Y-%m-%d") if m.date else "unknown date"
    line = f"- {date}: {m.home} {_fmt_score(m)} {m.away} ({m.competition}"
    if m.season:
        line += f" {m.season}"
    if m.round:
        line += f" Round {m.round}"
    if m.stage:
        line += f", {m.stage}"
    line += ")"
    return line


class SoccerQueries:
    """High-level queries over the loaded datasets."""

    def __init__(self, data: SoccerData):
        self.data = data

    # ------------------------------------------------------------------
    # match queries
    # ------------------------------------------------------------------
    def find_matches(
        self,
        team: Optional[str] = None,
        opponent: Optional[str] = None,
        competition: Optional[str] = None,
        season: Optional[int] = None,
        date_from: Optional[str] = None,
        date_to: Optional[str] = None,
        limit: int = 25,
    ) -> list[Match]:
        """Find matches by team, opponent, competition, season, date range."""
        d_from = datetime.strptime(date_from, "%Y-%m-%d") if date_from else None
        d_to = datetime.strptime(date_to, "%Y-%m-%d") if date_to else None

        out = []
        for m in self.data.competition_matches(competition=competition, season=season):
            if team and not (name_matches(team, m.home) or name_matches(team, m.away)):
                continue
            if opponent and not (
                name_matches(opponent, m.home) or name_matches(opponent, m.away)
            ):
                continue
            if team and opponent:
                # both teams must be playing each other
                t_home = name_matches(team, m.home) or name_matches(opponent, m.home)
                t_away = name_matches(team, m.away) or name_matches(opponent, m.away)
                if not (t_home and t_away):
                    continue
            if d_from and (m.date is None or m.date < d_from):
                continue
            if d_to and (m.date is None or m.date.date() > d_to.date()):
                continue
            out.append(m)
        out.sort(key=lambda m: (m.date or datetime.min))
        return out[:limit]

    def last_match_between(self, team_a: str, team_b: str) -> Optional[Match]:
        h2h = self.data.head_to_head(team_a, team_b)
        return h2h[-1] if h2h else None

    def format_match_list(self, matches: list[Match], total: Optional[int] = None) -> str:
        if not matches:
            return "No matches found."
        lines = [_match_line(m) for m in matches]
        total = total if total is not None else len(matches)
        if total > len(matches):
            lines.append(f"... ({total - len(matches)} more matches in dataset)")
        return "\n".join(lines)

    # ------------------------------------------------------------------
    # team queries
    # ------------------------------------------------------------------
    def team_stats(
        self,
        team: str,
        season: Optional[int] = None,
        competition: Optional[str] = None,
        venue: Optional[str] = None,  # 'home', 'away', or None (both)
    ) -> dict:
        """Win/loss/draw record and goals for a team, with optional filters."""
        stats = {
            "team": team,
            "season": season,
            "competition": competition,
            "venue": venue,
            "matches": 0,
            "wins": 0,
            "draws": 0,
            "losses": 0,
            "goals_for": 0,
            "goals_against": 0,
        }
        for m in self.data.competition_matches(competition=competition, season=season):
            is_home = name_matches(team, m.home)
            is_away = name_matches(team, m.away)
            if not (is_home or is_away):
                continue
            if venue == "home" and not is_home:
                continue
            if venue == "away" and not is_away:
                continue
            if m.home_goal is None or m.away_goal is None:
                continue
            stats["matches"] += 1
            if is_home:
                gf, ga = m.home_goal, m.away_goal
            else:
                gf, ga = m.away_goal, m.home_goal
            stats["goals_for"] += gf
            stats["goals_against"] += ga
            w = m.winner()
            if w == "draw":
                stats["draws"] += 1
            elif (w == "home") == is_home:
                stats["wins"] += 1
            else:
                stats["losses"] += 1
        if stats["matches"]:
            stats["win_rate"] = round(stats["wins"] / stats["matches"] * 100, 1)
        else:
            stats["win_rate"] = 0.0
        return stats

    def format_team_stats(self, stats: dict) -> str:
        bits = [stats["team"], "record"]
        if stats.get("season"):
            bits.append(f"({stats['season']}")
            if stats.get("competition"):
                bits.append(f"{stats['competition']}")
            bits.append(")")
        elif stats.get("competition"):
            bits.append(f"({stats['competition']})")
        if stats.get("venue"):
            bits.append(f"[{stats['venue']} matches]")
        header = " ".join(bits).replace(" )", ")")
        return (
            f"{header}:\n"
            f"- Matches: {stats['matches']}\n"
            f"- Wins: {stats['wins']}, Draws: {stats['draws']}, Losses: {stats['losses']}\n"
            f"- Goals For: {stats['goals_for']}, Goals Against: {stats['goals_against']}\n"
            f"- Win rate: {stats['win_rate']}%"
        )

    def head_to_head_summary(self, team_a: str, team_b: str) -> dict:
        summary = {
            "team_a": team_a,
            "team_b": team_b,
            "matches": 0,
            "team_a_wins": 0,
            "team_b_wins": 0,
            "draws": 0,
        }
        for m in self.data.head_to_head(team_a, team_b):
            if m.home_goal is None or m.away_goal is None:
                continue
            summary["matches"] += 1
            w = m.winner()
            if w == "draw":
                summary["draws"] += 1
            elif w == "home":
                if name_matches(team_a, m.home):
                    summary["team_a_wins"] += 1
                else:
                    summary["team_b_wins"] += 1
            else:
                if name_matches(team_a, m.away):
                    summary["team_a_wins"] += 1
                else:
                    summary["team_b_wins"] += 1
        return summary

    # ------------------------------------------------------------------
    # competition queries
    # ------------------------------------------------------------------
    def standings(self, competition: str, season: int) -> list[dict]:
        """Calculate a league table (3 points per win) from match results.

        Uses a single deduplicated source file per season so overlapping
        datasets (e.g. two files both covering the 2019 Brasileirão) do
        not double-count matches.
        """
        table: dict[str, dict] = {}

        def get(name: str) -> dict:
            key = normalize_name(name)
            if key not in table:
                table[key] = {
                    "team": name,
                    "matches": 0,
                    "wins": 0,
                    "draws": 0,
                    "losses": 0,
                    "goals_for": 0,
                    "goals_against": 0,
                    "points": 0,
                }
            return table[key]

        for m in self.data.competition_matches(competition=competition, season=season):
            if m.home_goal is None or m.away_goal is None:
                continue
            home, away = get(m.home), get(m.away)
            for side, gf, ga in (
                (home, m.home_goal, m.away_goal),
                (away, m.away_goal, m.home_goal),
            ):
                side["matches"] += 1
                side["goals_for"] += gf
                side["goals_against"] += ga
                if gf > ga:
                    side["wins"] += 1
                    side["points"] += 3
                elif gf == ga:
                    side["draws"] += 1
                    side["points"] += 1
                else:
                    side["losses"] += 1

        rows = sorted(
            table.values(),
            key=lambda r: (
                -r["points"],
                -(r["goals_for"] - r["goals_against"]),
                -r["goals_for"],
            ),
        )
        return rows

    def format_standings(self, rows: list[dict], competition: str, season: int) -> str:
        if not rows:
            return f"No standings data for {competition} {season}."
        lines = [f"{competition} {season} Standings (calculated from matches):"]
        for i, r in enumerate(rows, 1):
            gd = r["goals_for"] - r["goals_against"]
            tag = " - Champion" if i == 1 else ""
            lines.append(
                f"{i}. {r['team']} - {r['points']} pts "
                f"({r['wins']}W, {r['draws']}D, {r['losses']}L, GD {gd:+d}){tag}"
            )
        return "\n".join(lines)

    # ------------------------------------------------------------------
    # player queries
    # ------------------------------------------------------------------
    def search_players(
        self,
        name: Optional[str] = None,
        nationality: Optional[str] = None,
        club: Optional[str] = None,
        position: Optional[str] = None,
        order_by: str = "Overall",
        limit: int = 20,
    ) -> list[dict]:
        n = normalize_name(name) if name else None
        nat = normalize_name(nationality) if nationality else None
        cl = normalize_name(club) if club else None
        pos = position.strip().upper() if position else None
        out = []
        for p in self.data.players:
            if n and n not in (p.get("_norm_name") or ""):
                continue
            if nat and nat != normalize_name(p.get("Nationality") or ""):
                continue
            if cl and cl not in (p.get("_norm_club") or ""):
                continue
            if pos and (p.get("Position") or "").upper() != pos:
                continue
            out.append(p)
        out.sort(
            key=lambda p: p.get(order_by)
            if isinstance(p.get(order_by), (int, float))
            else 0,
            reverse=True,
        )
        return out[:limit]

    def format_players(self, players: list[dict]) -> str:
        if not players:
            return "No players found."
        lines = []
        for i, p in enumerate(players, 1):
            lines.append(
                f"{i}. {p.get('Name')} - Overall: {p.get('Overall')}, "
                f"Position: {p.get('Position')}, Club: {p.get('Club')}, "
                f"Nationality: {p.get('Nationality')}"
            )
        return "\n".join(lines)

    # ------------------------------------------------------------------
    # statistical analysis
    # ------------------------------------------------------------------
    def competition_stats(
        self, competition: Optional[str] = None, season: Optional[int] = None
    ) -> dict:
        played = home_wins = away_wins = draws = goals = 0
        for m in self.data.competition_matches(competition=competition, season=season):
            if m.home_goal is None or m.away_goal is None:
                continue
            played += 1
            goals += m.home_goal + m.away_goal
            w = m.winner()
            if w == "home":
                home_wins += 1
            elif w == "away":
                away_wins += 1
            else:
                draws += 1
        return {
            "competition": competition or "all competitions",
            "season": season,
            "matches": played,
            "total_goals": goals,
            "avg_goals_per_match": round(goals / played, 2) if played else 0.0,
            "home_wins": home_wins,
            "away_wins": away_wins,
            "draws": draws,
            "home_win_rate": round(home_wins / played * 100, 1) if played else 0.0,
        }

    def biggest_wins(
        self, competition: Optional[str] = None, limit: int = 10
    ) -> list[Match]:
        scored = []
        for m in self.data.competition_matches(competition=competition):
            if m.home_goal is None or m.away_goal is None:
                continue
            scored.append((abs(m.home_goal - m.away_goal), m.home_goal + m.away_goal, m))
        scored.sort(key=lambda t: (-t[0], -t[1]))
        return [m for _, _, m in scored[:limit]]

    def format_biggest_wins(self, matches: list[Match]) -> str:
        if not matches:
            return "No matches found."
        lines = ["Biggest victories in dataset:"]
        for i, m in enumerate(matches, 1):
            date = m.date.strftime("%Y-%m-%d") if m.date else "unknown date"
            lines.append(f"{i}. {date}: {m.home} {_fmt_score(m)} {m.away} ({m.competition})")
        return "\n".join(lines)

    def best_team_record(
        self, venue: str = "home", competition: Optional[str] = None, min_matches: int = 50
    ) -> list[dict]:
        """Rank teams by win rate at a venue ('home' or 'away')."""
        agg: dict[str, dict] = defaultdict(lambda: {"matches": 0, "wins": 0, "team": ""})
        for m in self.data.competition_matches(competition=competition):
            if m.home_goal is None or m.away_goal is None:
                continue
            if venue == "home":
                team, won = m.home, m.winner() == "home"
            else:
                team, won = m.away, m.winner() == "away"
            key = normalize_name(team)
            agg[key]["team"] = team
            agg[key]["matches"] += 1
            if won:
                agg[key]["wins"] += 1
        ranked = [
            {
                "team": v["team"],
                "matches": v["matches"],
                "wins": v["wins"],
                "win_rate": round(v["wins"] / v["matches"] * 100, 1),
            }
            for v in agg.values()
            if v["matches"] >= min_matches
        ]
        ranked.sort(key=lambda r: -r["win_rate"])
        return ranked

    def matches_per_team_summary(self, team: str) -> dict:
        """Per-competition breakdown of a team's matches."""
        per_comp: dict[str, int] = defaultdict(int)
        for m in self.data.team_matches(team):
            per_comp[m.competition] += 1
        return {"team": team, "competitions": dict(per_comp)}
